Add inference section to task guides (#18781)
huggingface/transformers: docs/source/en/tasks/asr.mdx

<Youtube id="TksaY_FDgnk"/>

> Automatic speech recognition (ASR) converts a speech signal to text, going from a sequence of audio inputs to textual outputs. Virtual assistants like Siri and Alexa use ASR models to help users everyday, and there are many other useful user-facing applications like live captioning and note-taking during meetings.

Reviewer: `textual` sounds a bit funky to me!

```suggestion
Automatic speech recognition (ASR) converts a speech signal to text, mapping a sequence of audio inputs to text outputs. Virtual assistants like Siri and Alexa use ASR models to help users everyday, and there are many other useful user-facing applications like live captioning and note-taking during meetings.
```

This guide will show you how to:

> 1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) to transcribe audio to text.

Reviewer: ```suggestion
1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset to transcribe audio to text.
```

2. Use your finetuned model for inference.

<Tip>

See the automatic speech recognition [task page](https://huggingface.co/tasks/automatic-speech-recognition) for more information about its associated models, datasets, and metrics.

</Tip>

Before you begin, make sure you have all the necessary libraries installed:

```bash

> pip install transformers datasets evaluate jiwer

Reviewer: Just checking that this example works without an audio processing library like `soundfile` or `librosa`? Great if it does! Some of these libraries have an annoying installation on Linux

Author: Yeah I was able to run the code without any issues :)

```

We encourage you to login to your Hugging Face account so you can upload and share your model with the community. When prompted, enter your token to login:

```py
>>> from huggingface_hub import notebook_login

>>> notebook_login()
```

## Load MInDS-14 dataset

> Start by loading a smaller subset of the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset from the 🤗 Datasets library. This'll give you a chance to experiment and make sure everythings works before spending more time training on the full dataset.

Reviewer: ```suggestion
Start by loading a smaller subset of the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset from the 🤗 Datasets library. This'll give you a chance to experiment and make sure everything works before spending more time training on the full dataset.
```

 'transcription': "hi I'm trying to use the banking app on my phone and currently my checking and savings account balance is not refreshing"}
```

Now create a preprocessing function that:

1. Calls the `audio` column to load and resample the audio file.
2. Extracts the `input_values` from the audio file and tokenize the `transcription` column with the processor.

```py
>>> def prepare_dataset(batch):
...     audio = batch["audio"]

> ...     batch = processor(audio["array"], sampling_rate=audio["sampling_rate"], text=batch["transcription"])

Reviewer: Nice!

 'transcription': "hi I'm trying to use the banking app on my phone and currently my checking and savings account balance is not refreshing"}
```

Now create a preprocessing function that:

1. Calls the `audio` column to load and resample the audio file.
2. Extracts the `input_values` from the audio file.
3. Tokenize the `transcription` column with the processor and return the `input_ids` as the labels.

```py
>>> def prepare_dataset(batch):
...     audio = batch["audio"]

...     batch["input_values"] = processor(audio["array"], sampling_rate=audio["sampling_rate"]).input_values[0]
...     batch["input_length"] = len(batch["input_values"])

> ...     batch["labels"] = processor(text=batch["transcription"]).input_ids

Author: Super cool! I tried passing it in one call, but I get a `KeyError: 'labels'` (I included a [notebook](https://colab.research.google.com/drive/1Sa0VU9GeABUieYjX2B9SUoK_Fh5JlkxS?usp=sharing)).

> ...     batch["labels"] = processor(text=batch["transcription"]).input_ids

Reviewer 2: The first line needs to be
```
batch = processor(audio["array"], sampling_rate=audio["sampling_rate"], text=batch["transcription"])
```
but then it worked for me.

> ...     batch["labels"] = processor(text=batch["transcription"]).input_ids

Author: Ok, this should be fixed now thanks to @sanchit-gandhi ! 👏

...         return batch
```

> Now create a batch of examples and dynamically pad them with `DataCollatorForCTCWithPadding`:

Reviewer: This is potentially a bit misleading since we don't actually batch or pad examples here, but simply instantiate the data collator

For a more in-depth example of how to fine-tune a model for automatic speech recognition, take a look at this blog [post](https://huggingface.co/blog/fine-tune-wav2vec2-english) for English ASR and this [post](https://huggingface.co/blog/fine-tune-xlsr-wav2vec2) for multilingual ASR.

</Tip>

## Inference

Great, now that you've finetuned a model, you can use it for inference!

Load an audio file you'd like to run inference on. Remember to resample the sampling rate of the audio file to match the sampling rate of the model if you need to!

```py
>>> from datasets import load_dataset, Audio

>>> dataset = load_dataset("PolyAI/minds14", "en-US", split="train")
>>> dataset = dataset.cast_column("audio", Audio(sampling_rate=16000))
>>> sampling_rate = dataset.features["audio"].sampling_rate
```

<frameworkcontent>

> <pt>

Reviewer 2: Why not use the pipeline function, at least at first? Then dive into processing + model.

Load a processor to preprocess the audio file and transcription, and return the `input` as PyTorch tensors (you can use the model you just finetuned or a pretrained model for automatic speech recognition):

```py
>>> from transformers import AutoProcessor

>>> processor = AutoProcessor.from_pretrained("facebook/wav2vec2-base-960h")
>>> inputs = processor(dataset[0]["audio"]["array"], sampling_rate=sampling_rate, return_tensors="pt")
```

Pass your inputs to the model and return the logits:

```py
>>> from transformers import AutoModelForCTC

> >>> model = AutoModelForCTC.from_pretrained("facebook/wav2vec2-base-960h")

Reviewer 2: This needs to be the fine-tuned model otherwise it defeats the point of this section.

</Tip>

## Inference

Great, now that you've finetuned a model, you can use it for inference!

Load an audio file you'd like to run inference on. Remember to resample the sampling rate of the audio file to match the sampling rate of the model if you need to!

```py
>>> from datasets import load_dataset, Audio

>>> dataset = load_dataset("PolyAI/minds14", "en-US", split="train")
>>> dataset = dataset.cast_column("audio", Audio(sampling_rate=16000))
>>> sampling_rate = dataset.features["audio"].sampling_rate
>>> audio_file = dataset[0]["audio"]["path"]
```

The simplest way to try out your finetuned model for inference is to use it in a [`pipeline`]. Instantiate a `pipeline` for automatic speech recognition with your model, and pass your audio file to it:

<Tip>

The results aren't very good because the model was only trained on a few examples to quickly demonstrate the process. Try finetuning your own model on the full dataset to get better results!

</Tip>

```py
>>> from transformers import pipeline

>>> transcriber = pipeline("automatic-speech-recognition", model="stevhliu/my_awesome_asr_minds_model")
>>> transcriber(audio_file)

> {'text': "ALDCBC</s>COCB</s>DCBDBCBRBCDCTCRWKDBK<s></s>KODT</s>C</s>T</s>C</s>OD<s>B</s>BCTOR</s>AKDOBODTCTEDQOATDVDKCDSKL<unk>KEKEPKNPKPADSRDSCLCQ<unk>ED<s>BDXAXDP<unk>KQDPWDKLEPRPKDTDRDPDNFRTADPLDSRDPDKDX<unk>XLXCXLCL</s>LCXLCLDT</s>TBLDLBLTLCB</s>BXTLTBLDTM</s>BLBLCLBLCLDT</s>LXCDTB</s>L</s>T</s>LTDK</s>TDTDTLOKDADPDSDT</s><unk>RDQTDDADSNDN</s>KD</s>T'DPNVPHKDPDRDRKRKD</s>PKDCQS<unk>X<unk>LTAQDXLXCTBCLBTQBCXBDQDTXM<unk>LCXL</s>CXT"}

Reviewer: Maybe we can train on a few hundred samples to get a somewhat reasonable transcription? Or is the point of this .md purely to demonstrate **how** to fine-tune (rather than with any care to performance), meaning we're happy with mumbo jumbo here?

Author: I think I may have run out of memory in Colab when I tried with more samples, but I'll try again with a few hundred samples to see if we can do better than this gibberish.

Reviewer: Note that when you download a dataset using `train[:20]` you actually download the entire dataset and then slice the first 20 samples -> so in terms of disk space using a couple of hundred samples will be equivalent to just 20.

For training, you might be getting OOM's because there are very long audio samples in your dataset. The self-attention mechanism memory scales with sequence length squared: $\text{memory} \propto \text{seq len}^2$. So there could be long samples > 20s that are causing OOM's! We usually filter such samples by a 20s max length criteria: 
https://github.com/huggingface/transformers/blob/a0f867430347bcf939f71d186409b9ca138c3b34/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L634-L642

Author: Finally trained a better model that doesn't return mumbo jumbo and added an important note on making sure the dataset text matches the casing of the tokenizer!

Reviewer: Nice! Glad to hear it's working 🗣
